docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+
+As an application developer I'd like to consume the TypeScript package published by the platform team and write application code that interacts with these resources (using whichever methods or permissions have been granted by the platform team). If I'm running my TypeScript code / container locally I expect the wing clients to connect to the wing simulator and interact with the simulated resources, and when my code is deployed to the AWS I expect the clients to connect to the real cloud resources.
+
+## Proposal
+
+The `clients` module has a `TypeScript` class used to create a client. When this is instantiated, `wing compile` will produce a publishable JavaScript npm package with TypeScript type declarations.
+
+```wing
+bring clients;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  // version: "0.0.0",
+  // outdir: join(@dirname, "dist"),
+);
+```
+
+To lift a resource into the client to make it accessible to consumers, add a `lift` expression:
+
+```wing
+bring cloud;
+
+let bucket = new cloud.Bucket();
+
+client.lift(lift { bucket: [put] });
+```
+
+In order to access the bucket, the client needs to know it should resolve the relevent information (e.g. bucket name or handle).
+This information is stored in client.context, so we need to way to both store and resolve this information.
+For example, we can store/retrieve the context from a file:
+
+```wing
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return fs.readJson("/var/run/wing-client-context.json");
+  },
+);
+
+// The consumer of the client will need to retrieve/link this data after deployment
+let clientContext = new cloud.Bucket() as "ctx";
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+If the runtime cost is acceptable, the context can even be directly resolved directly from the bucket during the client's runtime:
+
+```wing
+bring cloud;
+bring clients;
+
+let clientContext = new cloud.Bucket() as "ctx";
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return Json.parse(clientContext.get("ctx"));
+  },
+);
+
+// hydrates the context during deployment
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+In cases where reading from the filesystem is difficult and the overhead or a runtime lookup is too high, the context can be passed in via an environment variable:
+
+```wing
+bring cloud;
+bring clients;
+bring util;
+
+let bucket = new cloud.Bucket() as "shared";
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    let ctxRaw = util.getEnv("WING_CLIENT_CONTEXT");
+    let ctxDecoded = util.fromBase64(ctxRaw);
+    return Json.parse(ctxDecoded);
+  },
```

Comment:
I suspect custom resolvers will be common since the user will have to manually deal with the storing the context. I do like the idea of premade ones though.

Not sure this actually works right now, but WDYT about static inflights instead of an enum?

```wing
  contextResolver: clients.Resolver.FROM_ENV // <-- This is an inflight function
```

Reply:
Either sounds nice to me 👍